apps/server/src/modules/oauth-provider/service/oauth-provider.login-flow.service.ts

```diff
@@ -0,0 +1,61 @@
+import { AcceptLoginRequestBody, ProviderLoginResponse } from '@shared/infra/oauth-provider/dto';
+import { LoginRequestBody } from '@src/modules/oauth-provider/controller/dto';
+import { LtiToolDO } from '@shared/domain/domainobject/ltitool.do';
+import { LtiToolRepo, PseudonymsRepo, RoleRepo, UserRepo } from '@shared/repo';
+import { OauthProviderRequestMapper } from '@src/modules/oauth-provider/mapper/oauth-provider-request.mapper';
+import { ForbiddenException, NotFoundException } from '@nestjs/common';
+import { IResolvedUser, Permission, PermissionService, PseudonymDO, User } from '@shared/domain';
+import { ResolvedUserMapper } from '@src/modules/user/mapper';
+import { Injectable } from '@nestjs/common/decorators/core/injectable.decorator';
+
+@Injectable()
+export class OauthProviderLoginFlowService {
+	constructor(
+		private readonly ltiToolRepo: LtiToolRepo,
+		private readonly pseudonymsRepo: PseudonymsRepo,
+		private readonly userRepo: UserRepo,
+		private readonly roleRepo: RoleRepo,
+		private readonly permissionService: PermissionService
+	) {}
+
+	async setSubject(
+		currentUserId: string,
+		loginResponse: ProviderLoginResponse,
+		loginRequestBody: LoginRequestBody
+	): Promise<AcceptLoginRequestBody> {
+		if (loginResponse.client.client_id) {
+			const ltiToolDO: LtiToolDO = await this.ltiToolRepo.findByClientIdAndIsLocal(
+				loginResponse.client.client_id,
+				true
+			);
+			const pseudonym: PseudonymDO = await this.pseudonymsRepo.findByUserIdAndToolId(
+				currentUserId,
+				ltiToolDO.id as string
+			);
+			const acceptLoginRequestBody: AcceptLoginRequestBody = OauthProviderRequestMapper.mapCreateAcceptLoginRequestBody(
+				loginResponse,
+				loginRequestBody,
+				currentUserId,
+				pseudonym.pseudonym
+			);
+			return acceptLoginRequestBody;
+		}
+		throw new NotFoundException('Could not find oAuthClientId in login response to set subject');
+	}
+
+	async validateNextcloudPermission(currentUserId: string, loginResponse: ProviderLoginResponse) {
+		const user: User = await this.userRepo.findById(currentUserId, true);
+		const permissions: string[] = this.permissionService.resolvePermissions(user);
+		const resolvedUser: IResolvedUser = ResolvedUserMapper.mapToResponse(user, permissions, user.roles.getItems());
+
+		if (loginResponse.client.client_id) {
```

Comment:
that's the way it works right now in legacy

Reply:
please investigate if there is any reason for it to behave like that. This looks like a potenital security leak to me. Is there any reason not to pass a clientId? what semantics does it have if you do? how will the route respond?

Reply:
github is only showing me old code so I have to guess what you mean.
The hydra api does not state that the client_id parameter is always set in the login response, but we need it to identify, if the client is nextcloud. It is very unlikely that hydra will return a response without a client_id though, so it is more a check for the typescript compiler.
The check also uses hard-coded values and has to be solved another way later, but doing that is out of scope for porting the old code to nest.

Reply:
well if we have no clientID, then by definition we can not find a matching tool in our database.
So it would seem logical to me to behave as if we didnt find a matching tool, and throw an error.

Ideally either 500 or 502, with a hint like "expected hydra to return a clientId"